[JENKINS-71715] Allow plugins with alternate build log visualizations to handle console links from core and other plugins

jgreffe opened the pull request:
See [JENKINS-71715](https://issues.jenkins.io/browse/JENKINS-71715).

This PR introduces a new extension point `ConsoleUrlProvider`, which is intended to be used by Jelly files that link to the build console and plugins that redirect to the build console, in order for plugins which provide special build console visualizations (such as Pipeline: Graph View and Blue Ocean) to be able to serve their own visualizations in those contexts.

See downstream PRs here:
* An example of implementing `ConsoleUrlProvider`: https://github.com/jenkinsci/pipeline-graph-view-plugin/pull/289
* An example of using `Functions.getConsoleUrl` from a plugin: https://github.com/jenkinsci/workflow-job-plugin/pull/373
* An example of using `ConsoleUrlProvder.getRedirectUrl`: https://github.com/jenkinsci/pipeline-input-step-plugin/pull/145

I reviewed plugins in https://github.com/jenkinsci/bom for console links to see how they are affected by these changes, see https://github.com/jenkinsci/jenkins/pull/8321#issuecomment-1726449487.

### Testing done

New automated test added.

You can manually test the PR by running https://github.com/jenkinsci/pipeline-graph-view-plugin/pull/289 by itself. Note that you will have to configure "Console URL Provider" in either the global configuration or your user configuration for anything to  happen.

Here is a video of me testing [the downstream changes](https://github.com/jenkinsci/pipeline-graph-view-plugin/pull/289) with a Pipeline build:

https://github.com/jenkinsci/jenkins/assets/1068968/31a830f5-1012-4d99-bc80-8ed94e924078

### Proposed changelog entries

- Introduce an API for build visualization plugins to serve alternative build console views and an API for plugins to produce links to the build console.

### Proposed upgrade guidelines

N/A

### Submitter checklist
- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [x] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [x] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [x] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [x] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@dwnusbaum 

Before the changes are marked as `ready-for-merge`:

### Maintainer checklist
- [x] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [x] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [x] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [x] Proper changelog labels are set so that the changelog can be generated automatically.
- [x] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [x] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

jgreffe committed:
[JENKINS-71715] Allow plugins with alternate build log visualizations to handle console links from core and other plugins

jgreffe committed:
Merge branch 'master' into JENKINS-71715_first-approach

jgreffe committed:
Merge branch 'master' into JENKINS-71715_first-approach

dwnusbaum commented on core/src/main/java/hudson/link/ConsoleURLProvider.java:
```diff
@@ -0,0 +1,57 @@
+
+/**
+ * Extension point for providing console urls
+ * @since 2.417
```
I think the process is to leave these blank but in place for now, and then if/when this gets merged, someone will come back later after it gets released and update them.

```suggestion
 * @since TODO
```

jgreffe committed:
[JENKINS-71715] Apply suggestions with ordinal extension and @since TODO

jgreffe commented on core/src/main/java/hudson/link/ConsoleURLProvider.java:
```diff
@@ -0,0 +1,57 @@
+        return all().stream().findFirst().orElse(ConsoleURLProviderImpl.INSTANCE);
+    }
+
+    static class ConsoleURLProviderImpl extends ConsoleURLProvider {
```
Done with `@Extension(ordinal = -100)` as seen in some other classes.

jgreffe committed:
[JENKINS-71715] Public @Extension

daniel-beck commented on core/src/main/java/hudson/link/ConsoleURLProvider.java:
```diff
@@ -0,0 +1,59 @@
+        public String getRoot() {
+            String root = Jenkins.get().getRootUrl();
+            if (root == null) {
+                root = "http://unconfigured-jenkins-location/";
```
Is there a legitimate use case for when this code would be reached?

Display URL API has this because the URLs are used outside Jenkins and computed in a non-request context. It seems these URLs are always used in a request context, as they're used to render parts of the UI?

dwnusbaum replied:
Good point, it would probably be better to design the new API to work with URLs relative to the context root, and then use something like `'/' + Stapler.getCurrentRequest().getContextPath() + getConsoleURL(run)` to build the console URLs.

The implementation of the extension that should go into https://github.com/jenkinsci/display-url-api-plugin/pull/194 would need to convert the absolute URLs into context path-relative URLs though, and I guess what I was thinking [here](https://github.com/jenkinsci/pipeline-input-step-plugin/pull/145#discussion_r1281880000) would no longer apply, since we would want to use `ConsoleURLProvider` for all links in the Jenkins UI and `DisplayUrlProvider.getConsoleUrl` only when producing external links.

daniel-beck replied:
Thinking about it some more -- just after the ping comment, sorry Cyrille -- this could also easily be implemented by having a Jenkins-internal URL that redirects to the outside system, instead of linking there directly.

cyrille-leclerc replied:
Thanks @daniel-beck. Regarding the OpenTelemetry integrations where Jenkins pipeline logs are persisted in an external logs management system rather than in the Jenkins home, all integrations I am aware of want to use Jenkins's built-in logs visualization UX.
Those integrations

daniel-beck commented on core/src/main/java/hudson/link/ConsoleURLProvider.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    @Extension(ordinal = -100)
+    public static class ConsoleURLProviderImpl extends ConsoleURLProvider {
```
```suggestion
    public static class DefaultConsoleURLProviderImpl extends ConsoleURLProvider {
```
(Plus other required changes.)
To make the purpose clearer.

batmat committed:
Add or fix @since todo

we usually don't put versions but put todo instead and these versions get resolved _after_ merges.

Co-authored-by: Daniel Beck <1831569+daniel-beck@users.noreply.github.com>

jgreffe committed:
[JENKINS-71715] Apply suggestions from reviews

jglick commented:
Note to self that if merged, #7633 could take advantage of this API. Currently it just redirects to the index page of the new build, but it would be better to go straight to the log.

dwnusbaum commented on core/src/main/java/jenkins/link/ConsoleURLProvider.java:
```diff
@@ -0,0 +1,74 @@
+     * Defaults to {@link DefaultConsoleURLProviderImpl} otherwise.
+     */
+    public String getConsoleURL(Run<?, ?> run) {
+        return get().getConsoleURL(run);
```
This default implementation doesn't make sense as an instance method given how it is being called in `Run`. If not for `DefaultConsoleURLProviderImpl` it would recur infinitely. It should probably be `abstract` (probably better to make this class an interface though).

dwnusbaum commented on core/src/main/java/jenkins/link/ConsoleURLProvider.java:
```diff
@@ -0,0 +1,74 @@
+
+        @Override
+        public String getConsoleURL(Run<?, ?> run) {
+            return Stapler.getCurrentRequest().getContextPath() + "/" + Util.encode(run.getUrl()) + "console";
```
I think it would be nice to put the context path logic somewhere in the main class, perhaps a different static method that is the main API entrypoint, along with validation logic to make sure implementations return URLs with the expected structure (e.g. not absolute, leading/trailing slashes, whatever).

We could do something like `Functions.getActionUrl` and allow different types of URLs, but IDK if we want to do that or be more restrictive:
  https://github.com/jenkinsci/jenkins/blob/eb96b8e07bcf2da498411ede238d24a18ec075c3/core/src/main/java/hudson/Functions.java#L1888-L1904

jgreffe committed:
[JENKINS-71715] Switch ConsoleURLProvider to interface and add Functions.getContextRelativeUrl

daniel-beck commented on core/src/main/java/jenkins/link/ConsoleURLProvider.java:
```diff
@@ -27,48 +27,40 @@
      */
-    public static ConsoleURLProvider get() {
+    static ConsoleURLProvider get() {
         return all().stream().findFirst().orElseThrow(() -> new RuntimeException("The DefaultConsoleURLProviderImpl should be instanciated"));
```
```suggestion
        return all().stream().findFirst().orElseThrow(() -> new IllegalStateException("No ConsoleURLProvider exists"));
```

daniel-beck reviewed (commented):
How would this be used by an implementation? Given this looks at the highest priority implementation in the static extension list, do admins need https://plugins.jenkins.io/extension-filter/ to configure that? Or is the expectation really that nothing but `display-url-api` implements this? The answer should be documented in Javadoc.

dwnusbaum commented on core/src/main/java/hudson/Functions.java:
No worries, I'm going to delete this.

dwnusbaum replied:
See https://github.com/jenkinsci/jenkins/pull/8321/commits/d8530041874cc01c9082089a77c65651a1a7a149. I deleted the generic utility method and replaced it with one that is intended to be used by Jelly files that want to produce build console links.

dwnusbaum commented:
> How would this be used by an implementation? Given this looks at the highest priority implementation in the static extension list, do admins need https://plugins.jenkins.io/extension-filter/ to configure that? Or is the expectation really that nothing but display-url-api implements this? 

I am going to change things. I think I'll swap to looping over all extensions, and allowing implementations to return `null` if they don't want to do anything for a particular `run`. Thinking about it some more, it might be better to only have actual visualization plugins implement the extension point if desired and leave `display-url-api` for external URLs only.

dwnusbaum commented on core/src/main/java/hudson/link/ConsoleURLProvider.java:
```diff
@@ -0,0 +1,59 @@
+package hudson.link;
```
Is it better to create a `jenkins.console` package and put it there or put it in the already-existing `hudson.console` package?

dwnusbaum committed:
[JENKINS-71715] Refactor ConsoleUrlProvider API and add Functions.getConsoleUrl method for use in Jelly files

dwnusbaum commented on core/src/main/java/jenkins/console/ConsoleURLProvider.java:
```diff
@@ -0,0 +1,90 @@
+     * @param run the build
+     * @return the URL for the console for the specified build, relative to the web server root
+     */
+    static @NonNull String getRedirectUrl(Run<?, ?> run) {
```
This method is intended to be used directly in cases where Java code wants to redirect a request to the console as in https://github.com/jenkinsci/pipeline-input-step-plugin/pull/145.

dwnusbaum commented on core/src/main/resources/hudson/model/AbstractBuild/changes.jelly:
```diff
@@ -40,7 +40,7 @@ THE SOFTWARE.
         </j:when>
         <j:otherwise>
-          ${%Failed to determine} (<a href="console">${%log}</a>)
+          ${%Failed to determine} (<a href="${h.getConsoleUrl(it)}">${%log}</a>)
```
FTR: I have not tested all of these changes yet.

dwnusbaum replied:
Ok, I believe I have checked all of these links manually now.

dwnusbaum commented on core/src/main/resources/lib/hudson/buildProgressBar.jelly:
```diff
@@ -26,6 +26,7 @@ THE SOFTWARE.
   <st:documentation>
     Progress bar for a build in progress.
 
+    <!-- TODO: Do Functions.getConsoleUrl and ConsoleUrlProvider need to be adapted to use Queue.Executable instead of Run? -->
```
I think making that change would be fine, but I will investigate this case to understand if it is necessary.

dwnusbaum replied:
I did some quick testing with `node` steps and Pipeline and `PlaceholderExecutable` does seem to hit this case. I think we can probably special-case this particular call (either here or in `Functions.getConsoleUrl`) rather than needing to change the API itself though.

dwnusbaum replied:
Well after some very annoying debugging I finally realized that the `node` step defines its own `executorCell.jelly`, which actually misuses this taglib and passes a `Queue.Task` (`PlaceholderTask`) instead of `Queue.Executable` (`PlaceholderExecutable`), see https://github.com/jenkinsci/workflow-durable-task-step-plugin/blob/94bdc2b50c6fbab37526804d232eb2f1da971db9/src/main/resources/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution/PlaceholderTask/PlaceholderExecutable/executorCell.jelly#L35. So, to fix this here in core as-is, we'd have to make `Functions.getConsoleUrl` take `Object`, and then extract the `Run` from either a `Queue.Task` or `Queue.Executable`.

dwnusbaum replied:
I think I'll file a PR against `workflow-durable-task-step` though and just handle `Queue.Executable` here as documented.

dwnusbaum replied:
See https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/330.

dwnusbaum replied:
Addressed in [46b9cda](https://github.com/jenkinsci/jenkins/pull/8321/commits/46b9cda076f1184c846dc4c816c8568982719de0).

dwnusbaum commented on core/src/main/resources/lib/hudson/project/console-link.jelly:
```diff
@@ -28,11 +28,11 @@ THE SOFTWARE.
         <j:when test="${it.logFile.length() > 200000}">
             <!-- Show raw link directly so user need not click through live console page, though this is not so bad now as they would just see: Skipping nnn KB.. Full Log. -->
-            <l:task href="${buildUrl.baseUrl}/console" icon="icon-terminal icon-md" title="${%Console Output}"/>
+            <l:task href="${h.getConsoleUrl(it)}" icon="icon-terminal icon-md" title="${%Console Output}"/>
             <l:task href="${buildUrl.baseUrl}/consoleText" icon="icon-document icon-md" title="${%View as plain text}"/>
```
I am not entirely sure what we want to do about this case. I think it makes sense to redirect the main "console output" link, but IDK if we should keep the "View as plain text" link or hide it if the console URL is non-default or something like that. For now I am inclined to just leave it as-is.

dwnusbaum commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -1903,6 +1904,16 @@ public static String joinPath(String... components) {
+     * @return the absolute URL for accessing the specified build's console
+     * @since TODO
+     */
+    public static @CheckForNull String getConsoleUrl(Run<?, ?> run) {
```
This method is intended for use in Jelly files (both in core and in plugins). I think it is a little cleaner to put it here rather than directly on `Run`, but that is just my opinion.

dwnusbaum commented on core/src/main/java/jenkins/console/ConsoleURLProvider.java:
```diff
@@ -0,0 +1,90 @@
+     */
+    static @NonNull String getRedirectUrl(Run<?, ?> run) {
+        String url = null;
+        for (ConsoleUrlProvider provider : ExtensionList.lookup(ConsoleUrlProvider.class)) {
```
@daniel-beck raised a few important points in https://github.com/jenkinsci/jenkins/pull/8321#pullrequestreview-1563282385 and I am not entirely sure how we should address them.

With my latest changes to the API, we can create implementations that only care about Pipelines, or builds with odd numbers, etc. However, if both Blue Ocean and Pipeline Graph View wanted to implement `ConsoleUrlProvider` for all Pipeline builds, and you have both installed on a single controller, they would compete for precedence based on `ordinal`. How do we want to address this? I can see a few approaches:
* Do nothing, let admins use https://plugins.jenkins.io/extension-filter/ as needed. Maybe this is good enough if we only expect a few overlapping implementations of this extension point, especially if we think the implementations are likely to come from competing plugins where users are not likely to them at the same time.
* Provide a `GlobalConfiguration` UI in core to allow admins to select which `ConsoleUrlProvider` extensions should be enabled on a given controller, or allow them to be sorted based on the admin's preference. Same end result as above, though arguably more user-friendly, but also it has additional implementation complexity.
* Provide a user-based configuration in core to let users select which `ConsoleUrlProvider` extensions they want to allow, or allow them to be sorted based on the user's preference. Adds even more implementation complexity than the above options, but now if there are two plugins installed with competing `ConsoleUrlProvider` implementations, individual users get to pick which one they want to use.

For now at least I need to update https://github.com/jenkinsci/pipeline-graph-view-plugin/pull/288 and https://github.com/jenkinsci/pipeline-input-step-plugin/pull/145 based on the latest changes.

timja replied:
I suspect you will want 2 and 3, admins to set a default and allow users to override it

daniel-beck replied:
Missed this thread and posted https://github.com/jenkinsci/jenkins/pull/8321#issuecomment-1666464680. For now IMO option 2 should be enough (option 3 only if sufficient user demand exists).

timja replied:
I feel you'll get users that are used to it but some users who would want the other implementation, also with the display url only having a per user setting it seems weird for this one not to?

timja replied:
For a cloudbees plugin it may make more sense as a control of the whole instance.

In the case of pipeline graph view plugin someone may want to fallback to the default console view if it doesn't handle something for them or they have used it for 10+ years and do not want to change.

also PGV doesn't support freestyle builds at all and I haven't seen any handling for that in the downstream PR.

dwnusbaum replied:
> also PGV doesn't support freestyle builds at all and I haven't seen any handling for that in the downstream PR.

Yes sorry, all of the downstream PRs need to be updated. I will get to that later today.

dwnusbaum replied:
See [018789a](https://github.com/jenkinsci/jenkins/pull/8321/commits/018789a84712f3f112b69dae6d0739325a26adcb) and let me know what you think. We can dial it back as desired.

timja replied:
this looks good to me

daniel-beck commented:
> I think I'll swap to looping over all extensions, and then allowing implementations to return `null` if they don't want to do anything for a particular `run`

I'm wondering what the configuration would look like then if you end up with two implementations. Find out whichever is the higher precedence (how?) and disable that in a plugin-defined manner, so it falls through to the other, seems like an awkward admin experience.

A global configuration that only shows up when you have a non-default implementation (similar to Artifact Management for Builds) might be a reasonable UI for this: Admins add/remove/sort descriptors of what's not the default. Could even gracefully reduce to an enable/disable checkbox if there's just one implementation.

dwnusbaum committed:
[JENKINS-71715] Fix casing of ConsoleURLProvider.java

dwnusbaum committed:
[JENKINS-71715] Add global and per-user configuration options for ConsoleUrlProvider

dwnusbaum committed:
[JENKINS-71715] Fix Javadoc reference

dwnusbaum committed:
[JENKINS-71715] Remove outdated TODO comment

dwnusbaum committed:
Merge branch 'master' into JENKINS-71715_first-approach

dwnusbaum commented on core/src/main/java/jenkins/console/ConsoleUrlProvider.java:
```diff
@@ -0,0 +1,130 @@
+        ConsoleUrlProviderGlobalConfiguration globalConfig = ConsoleUrlProviderGlobalConfiguration.get();
+        List<ConsoleUrlProvider> providers = globalConfig.getProviders();
+        User currentUser = User.current();
+        if (currentUser != null) { // TODO: Should we add a configuration option to allow admins to suppress this behavior?
```
In other words, do admins want to be able to restrict users from being able to configure things, or do they just want to set defaults? For now admins could use extension-filter on `ConsoleUrlProviderUserProperty.DescriptorImpl` if they really want full control, so I am not inclined to add additional options.

dwnusbaum replied:
Given no one seems to think this is necessary, I dropped the comment for now.

timja commented on core/src/main/java/jenkins/console/ConsoleUrlProvider.java:
```diff
@@ -0,0 +1,130 @@
+                }
+            }
+        }
+        if (url == null) {
```
is this ever hit? won't the default one be hit if there aren't any others?

dwnusbaum replied:
This gets hit if admins never configure anything, or if they do not add the default extension to the global configuration. The default extension is really only useful for individual users in case they do not want to use whatever their admin selects as the default, but it seemed complicated to try to filter out the default extension only in the global case. I can look into that if desired though.

dwnusbaum replied:
Explanatory comment added in [5f051c9](https://github.com/jenkinsci/jenkins/pull/8321/commits/5f051c9b0faf10e39f48122389f38b05ce2b3ef6).

timja commented on core/src/main/java/jenkins/console/ConsoleUrlProviderGlobalConfiguration.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    public boolean isEnabled() {
+        return !Jenkins.get().getDescriptorList(ConsoleUrlProvider.class).isEmpty();
```
won't there always be at least one? shouldn't it be if the only 1 isn't the default one?

dwnusbaum replied:
Yes good catch, this was before I added the default extension. I'll update it.

dwnusbaum replied:
Fixed in [5f051c9](https://github.com/jenkinsci/jenkins/pull/8321/commits/5f051c9b0faf10e39f48122389f38b05ce2b3ef6).

timja commented on core/src/main/resources/jenkins/console/ConsoleUrlProviderGlobalConfiguration/config.jelly:
```diff
@@ -0,0 +1,13 @@
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:f="/lib/form">
+    <j:if test="${h.hasPermission(app.ADMINISTER) and descriptor.enabled}">
```
why is this permission check needed it's a global configuration value?

This prevents System read users from even seeing the config which seems wrong

dwnusbaum replied:
Copied from elsewhere. I will check this.

dwnusbaum replied:
Removed in [5f051c9](https://github.com/jenkinsci/jenkins/pull/8321/commits/5f051c9b0faf10e39f48122389f38b05ce2b3ef6).

timja reviewed (commented):
not tested locally yet, some initial feedback

dwnusbaum committed:
[JENKINS-71715] Miscellaneous updates based on review feedback

dwnusbaum committed:
[JENKINS-71715] Fix Javadoc links

dwnusbaum committed:
[JENKINS-71715] Update console links for build list and build time trend widgets

dwnusbaum commented on core/src/main/java/jenkins/console/ConsoleUrlProvider.java:
```diff
@@ -0,0 +1,121 @@
+            url = run.getUrl() + "console";
+        }
+        // TODO:
+        // * Fail if absolute URL?
```
e.g., if an extension returns `https://www.example.com`, do we to support that case (skip context path stuff below), explicitly fail by throwing an exception, implicitly fail by building an invalid URL, etc.

daniel-beck replied:
I'd treat it like a `null` URL returned from a provider, invalid return value.

dwnusbaum replied:
Done in [b5420ad](https://github.com/jenkinsci/jenkins/pull/8321/commits/b5420ad07c46402632bdd712a5cac7ffb88bda04). I also added a test for the new logic.

dwnusbaum committed:
[JENKINS-71715] Functions.getConsoleUrl must take Queue.Executable instead of Run

dwnusbaum commented:
Ok, just for the record, I think have tested all of the modified URLs in this PR. I would like to get a bit more feedback to see if everyone is ok with the new API and this approach in general, and if so, I will then look through all plugins in https://github.com/jenkinsci/bom and file draft PRs to update any console links to use the new APIs.

dwnusbaum committed:
[JENKINS-71715] Simplify test using WebClient.executeOnServer

dwnusbaum committed:
Merge branch 'master' into JENKINS-71715_first-approach

jgreffe reviewed (commented):
LGTM.

dwnusbaum committed:
Merge branch 'master' into JENKINS-71715_first-approach

jgreffe committed:
[JENKINS-71715] Add constructor

dwnusbaum committed:
Merge branch 'master' into JENKINS-71715_first-approach

jgreffe committed:
Merge branch 'master' into JENKINS-71715_first-approach

dwnusbaum committed:
Merge branch 'master' into JENKINS-71715_first-approach

dwnusbaum commented:
@jenkinsci/core-pr-reviewers Any thoughts on this PR, in particular whether you are generally ok with the idea and approach?

timja commented:
> @jenkinsci/core-pr-reviewers Any thoughts on this PR, in particular whether you are generally ok with the idea and approach?

fine with the idea and approach I just haven't had time to test it out

dwnusbaum committed:
Merge branch 'master' into JENKINS-71715_first-approach

dwnusbaum commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -1903,6 +1905,23 @@ public static String joinPath(String... components) {
+    public static @CheckForNull String getConsoleUrl(Queue.Executable executable) {
+        if (executable == null) {
+            return null;
+        } else if (executable instanceof Run) {
+            return ConsoleUrlProvider.getRedirectUrl((Run<?, ?>) executable);
```
I was thinking about this, and I could see an argument for changing `ConsoleUrlProvider` to work with `Queue.Executable` directly as well. For example, branch indexing and org folder scans use `FolderComputation`, which is a `Queue.Executable` but has no associated `Run`. I am somewhat doubtful that anyone will want to create external visualizations for those logs, but maybe if we are introducing a new API anyway we should make it as generic as possible? The tradeoff is additional complexity in each individual extension, but maybe we could add a `ConsoleUrlProvider.ForRun` subclass that most implementations could use and only handles `Run`, and then anyone who really needs to do something special for `Queue.Executable` could implement `ConsoleUrlProvider` directly?

dwnusbaum commented:
I went ahead and searched plugins in `jenkinsci/bom` for `console` using Github and did not find any other plugins that produce in-UI links to build consoles. Here are the interesting cases I found:
* [`email-ext`](https://github.com/search?q=repo%3Ajenkinsci%2Femail-ext-plugin%20console&type=code) has some standard templates that link to `/console`. These could and should not not use the new API, they would need something more like https://github.com/jenkinsci/display-url-api-plugin/pull/194. We dropped that PR because we didn't have any known consumers for the new API, so finding this is somewhat interesting and could be used to reconsider that PR.
* [`cloudbees-folder`](https://github.com/search?q=repo%3Ajenkinsci%2Fcloudbees-folder-plugin%20console&type=code) produces links to `<ComputedFolder>/console` for folder computation logs, which is related to https://github.com/jenkinsci/jenkins/pull/8321#discussion_r1327656268. Unless we make those changes, there is no reason to change those links in the plugin for now. That said, it also has [an interesting use of `<p:console-link>`](https://github.com/jenkinsci/cloudbees-folder-plugin/blob/e3bfd7839a815d89d3580dcf41745cf2d5cb4b14/src/main/resources/com/cloudbees/hudson/plugins/folder/computed/PseudoRun/sidepanel.jelly#L32) (which is edited in this PR) for a [`PseudoRun`](https://github.com/jenkinsci/cloudbees-folder-plugin/blob/e3bfd7839a815d89d3580dcf41745cf2d5cb4b14/src/main/java/com/cloudbees/hudson/plugins/folder/computed/PseudoRun.java) class that is not a `Run` or a `Queue.Executable` ([it just has special handlers for `doConsole` and `doConsoleText`](https://github.com/jenkinsci/cloudbees-folder-plugin/blob/e3bfd7839a815d89d3580dcf41745cf2d5cb4b14/src/main/java/com/cloudbees/hudson/plugins/folder/computed/PseudoRun.java#L111-L117)), so I need to check whether that Jelly tag still works in that context with this PR.
* [`branch-api`](https://github.com/search?q=repo%3Ajenkinsci%2Fbranch-api-plugin%20console&type=code) produces links to branch indexing logs for builds triggered by branch indexing, which is similar to the first point regarding `cloudbees-folder` and only needs to be changed if https://github.com/jenkinsci/jenkins/pull/8321#discussion_r1327656268 is implemented.

dwnusbaum commented:
> That said, it also has [an interesting use of <p:console-link>](https://github.com/jenkinsci/cloudbees-folder-plugin/blob/e3bfd7839a815d89d3580dcf41745cf2d5cb4b14/src/main/resources/com/cloudbees/hudson/plugins/folder/computed/PseudoRun/sidepanel.jelly#L32) (which is edited in this PR) for a [PseudoRun](https://github.com/jenkinsci/cloudbees-folder-plugin/blob/e3bfd7839a815d89d3580dcf41745cf2d5cb4b14/src/main/java/com/cloudbees/hudson/plugins/folder/computed/PseudoRun.java) class that is not a Run or a Queue.Executable ([it just has special handlers for doConsole and doConsoleText](https://github.com/jenkinsci/cloudbees-folder-plugin/blob/e3bfd7839a815d89d3580dcf41745cf2d5cb4b14/src/main/java/com/cloudbees/hudson/plugins/folder/computed/PseudoRun.java#L111-L117)), so I need to check whether that Jelly tag still works in that context with this PR.

I checked and indeed the changes to `p:console-link` undesirably suppress a link to the folder computation logs. Unfortunately I see no clear way to simplify things because `PseudoRun` does not implement any types that would make sense for `ConsoleUrlProvider` to be based on, so I think we will just need to preserve the old behavior as a fallback for this taglib to keep that link working for `cloudbees-folder`.

dwnusbaum committed:
Merge branch 'master' into JENKINS-71715_first-approach

dwnusbaum committed:
[JENKINS-71715] Preserve old behavior of p:console-link as a fallback for atypical consumers

daniel-beck commented:
Noting https://github.com/jenkinsci/jenkins/pull/8403 which is probably relevant for this PR.

daniel-beck commented on core/src/main/java/jenkins/console/ConsoleUrlProvider.java:
```diff
@@ -0,0 +1,121 @@
+            ConsoleUrlProviderUserProperty userProperty = currentUser.getProperty(ConsoleUrlProviderUserProperty.class);
+            if (userProperty != null) {
+                List<ConsoleUrlProvider> userProviders = userProperty.getProviders();
+                if (userProviders != null) {
+                    providers = userProviders;
+                }
```
This looks like a user who configures a provider that does return `null` will not fall back to the admin choice, but to the default provider. Is that intentional?

dwnusbaum replied:
I updated the behavior and added a test in [b5420ad](https://github.com/jenkinsci/jenkins/pull/8321/commits/b5420ad07c46402632bdd712a5cac7ffb88bda04).

daniel-beck commented on core/src/main/java/jenkins/console/ConsoleUrlProviderGlobalConfiguration.java:
```diff
@@ -0,0 +1,90 @@
+
+    @Override
+    public boolean configure(StaplerRequest req, JSONObject json) throws FormException {
+        // We have to null out providers before data binding to allow all providers to be deleted in the config UI.
```
Unsure what this means. Is this about a `null` value vs. an empty list?

dwnusbaum commented on core/src/main/java/jenkins/console/ConsoleUrlProvider.java:
```diff
@@ -0,0 +1,121 @@
+ * @since TODO
+ */
+public interface ConsoleUrlProvider extends Describable<ConsoleUrlProvider> {
+    @Restricted(NoExternalUse.class)
+    Logger LOGGER = Logger.getLogger(ConsoleUrlProvider.class.getName());
```
Since `ConsoleUrlProvider` is an interface, the field can't be `private` and it is `public static final` by default. I think we could use a dynamic `private Logger getLogger()` method that calls `Logger.getLogger` each time to avoid the field though if preferred (not sure how costly it is to call `Logger.getLogger` on demand).

dwnusbaum committed:
[JENKINS-71715] Ignore invalid URLs and consult global providers if user providers fail to provide a valid URL

dwnusbaum commented:
> Noting https://github.com/jenkinsci/jenkins/pull/8403 which is probably relevant for this PR.

Ok, should I go ahead and move the new global configuration from this PR into that category? I think it is conceptually very similar to `display-url-api`, which was mentioned in that PR:

> Could possibly move the Display notification URL provider to appearance too? (the thing that selects Classic, Blue Ocean or Pipeline Graph View)

timja commented:
Yes please

dwnusbaum committed:
[JENKINS-71715] Move ConsoleUrlProviderGlobalConfiguration into AppearanceCategory

daniel-beck commented:
It's not that easy though, otherwise we get

> <img width="711" alt="Screenshot 2023-09-28 at 09 21 36" src="https://github.com/jenkinsci/jenkins/assets/1831569/c66104b1-8bea-4a55-af58-cffa791c8f93">

Since both category and this extension point are a core feature, this PR could add special behavior for this case in https://github.com/jenkinsci/jenkins/blob/9dabd874c2b183339ac419ea94b637db8f206b62/core/src/main/java/jenkins/appearance/AppearanceGlobalConfiguration.java#L56-L57

dwnusbaum committed:
[JENKINS-71715] Hide AppearanceGlobalConfiguration if ConsoleUrlProviderGlobalConfiguration is disabled and would be the only option on the page

jeromepochat commented:
I just noticed that "Appearance" page is still reachable through direct URL `http://.../manage/appearance/` in case of single `ConsoleUrlProvider` available. Could the link be enabled or disabled similarly to action button (depending on number of available providers)?

dwnusbaum commented:
> Could the link be enabled or disabled similarly to action button (depending on number of available providers)?

I'm not 100% sure, but I don't think there is a simple way to do it. That behavior should be the same though with or without this PR: if you install the latest Jenkins weekly, do not install any plugins, and go to `manage/appearance/`, you will see a blank page.

jeromepochat commented:
> I'm not 100% sure, but I don't think there is a simple way to do it. That behavior should be the same though with or without this PR: if you install the latest Jenkins weekly, do not install any plugins, and go to `manage/appearance/`, you will see a mostly blank page.

:+1:

I wrongly thought the "Appearance" page was introduced by this PR.

Then everything looks good to me.

fcojfernandez reviewed (approved):
Nice feature

jglick commented on core/src/main/resources/jenkins/console/ConsoleUrlProviderUserProperty/config.jelly:
```diff
@@ -0,0 +1,15 @@
+<j:jelly xmlns:j="jelly:core" xmlns:f="/lib/form">
+    <j:if test="${descriptor.enabled}">
+        <f:invisibleEntry>
+          <!-- Dummy field to always have an entry in form submissions so users can delete all providers. -->
```
Do not do it this way. Rather start `doConfigSubmit` or whatever by nulling out the field before binding properties.

dwnusbaum replied:
IIRC, both were needed, otherwise `configure` didn't even get called. I will check again though, perhaps this can be deleted.

dwnusbaum replied:
Ah yeah I think this behavior was specific to user properties actually. I'll still check it again though.

jglick replied:
Probably what you ran into is that form binding only provides a JSON element for a list if it has at least one entry. So for a top-level (reconfigurable) object, if you just use the default `bindJSON`, after deleting all elements of a list and saving, no change is applied to that property. The workaround is to simply null out any list-valued fields before calling `bindJSON`.

dwnusbaum replied:
For `ConsoleUrlProviderGlobalConfiguration`, yes, that was a problem and that is how I fixed it.

`ConsoleUrlProviderUserProperty` is different though. [`User.doConfigSubmit`](https://github.com/jenkinsci/jenkins/blob/c9947a3473adbb2f3242d5facb37cd82bd801675/core/src/main/java/hudson/model/User.java#L867) is unusual and has complex logic regarding creating new properties vs reconfiguring existing ones. IIRC, in the problematic case, since `ConsoleUrlProviderUserProperty` only has a single `List` field, [this condition](https://github.com/jenkinsci/jenkins/blob/c9947a3473adbb2f3242d5facb37cd82bd801675/core/src/main/java/hudson/model/User.java#L881) is `true` because the property itself is not present in the submitted JSON at all, and so the logic there chooses to reuse the existing property if one exists.

Whether that logic was intentional was not clear to me. Perhaps [these lines](https://github.com/jenkinsci/jenkins/blob/c9947a3473adbb2f3242d5facb37cd82bd801675/core/src/main/java/hudson/model/User.java#L889-L892) should be moved inside of [this block](https://github.com/jenkinsci/jenkins/blob/c9947a3473adbb2f3242d5facb37cd82bd801675/core/src/main/java/hudson/model/User.java#L881-L887), so that any properties not present in the submission are deleted, but I have no idea if anything relies on the current behavior (perhaps this case is intended to preserve user properties that do not offer any GUI configuration?) and it was not something I wanted to look into here.

dwnusbaum replied:
Hmm, maybe actually I can reuse the lock inside of `TimingFlowNodeStorage` instead to simplify things.

jglick replied:
> perhaps this case is intended to preserve user properties that do not offer any GUI configuration?

Perhaps. (Or those whose GUI configuration is on a separate page.)

> https://github.com/jenkinsci/jenkins/blob/c9947a3473adbb2f3242d5facb37cd82bd801675/core/src/main/java/hudson/model/User.java#L881 is `true` because the property itself is not present in the submitted JSON at all

`false` you mean? Smells like a bug in the form submission JSON generator; it should at least create `{}` maybe with `_class`.

dwnusbaum replied:
Yes sorry I meant `false`.

daniel-beck commented on core/src/main/resources/jenkins/console/ConsoleUrlProviderUserProperty/config.properties:
```diff
@@ -0,0 +1 @@
+description=Controls what visualization tool handles build console links. Providers higher in the list take precedence over providers lower in the list.
```
Ideally this would explain the relationship to the global configuration by the admin (precedence / override).

dwnusbaum replied:
Let me know what you think about [64c3faa](https://github.com/jenkinsci/jenkins/pull/8321/commits/64c3faa8ac16742878377db5f9b0d14ff95a1338).

daniel-beck commented on core/src/main/resources/jenkins/console/ConsoleUrlProviderGlobalConfiguration/config.properties:
```diff
@@ -0,0 +1 @@
+description=Controls what visualization tool handles build console links. Providers higher in the list take precedence over providers lower in the list.
```
```suggestion
description=Controls what visualization tool handles build console links. Providers higher in the list take precedence over providers lower in the list. If no explicitly specified provider can render a console link, the default provider will be used.
```

dwnusbaum replied:
Added in [64c3faa](https://github.com/jenkinsci/jenkins/pull/8321/commits/64c3faa8ac16742878377db5f9b0d14ff95a1338), along with some other text.

daniel-beck commented on core/src/main/resources/jenkins/console/ConsoleUrlProviderGlobalConfiguration/config.jelly:
```diff
@@ -0,0 +1,13 @@
+                ${%description}
+            </f:block>
+            <f:entry>
+                <f:repeatableHeteroProperty field="providers" hasHeader="true" oneEach="true" header="${%Providers}" />
```
It's a bit awkward that the default appears in the global list. I suggest this be filtered to not show Default explicitly (by implementing `#getProvidersDescriptors` AFAICT).

dwnusbaum replied:
Nice, thanks for the pointer to the expected pattern for the method name. Done in [64c3faa](https://github.com/jenkinsci/jenkins/pull/8321/commits/64c3faa8ac16742878377db5f9b0d14ff95a1338).

daniel-beck reviewed (approved):
Works fine in testing with the downstream PRs. Code and tests look reasonable as well.

Some (optional) thoughts:

* It'd be useful to have developer guidelines for implementations. For example, implementing plugins no longer need a separate link to their own console (as the current PR of `pipeline-graph-view` still does), but OTOH would be good for them to a link to the legacy build log in their custom UI.
* The current wording focuses on the URLs/links because that's what it does from a technical POV. From a user POV, it seems to me it's more about the chosen build log visualization. Perhaps someone more in the PM/PO/UX space can provide feedback on the name of this?

dwnusbaum commented:
> The current wording focuses on the URLs/links because that's what it does from a technical POV. From a user POV, it seems to me it's more about the chosen build log visualization

Sure, we could change "Console URL Provider" to "Console Visualization Provider" at least in the user-facing text, and also the class names if desired.

dwnusbaum commented:
I will also add a note for implementors to the `ConsoleUrlProvider` Javadoc regarding some of the points you mentioned.

> implementing plugins no longer need a separate link to their own console

I don't think this is so clear-cut. The console visualization plugins cannot trivially tell whether the current user has their visualization selected to know if the link is redundant (they could consult `ConsoleUrlProvider.getRedirectUrl` but it would be somewhat awkward), and even if a user prefers one visualization, they may still want to be able to easily access other visualizations. It seems simplest to me to just leave the redundant links in place in the sidepanel of the build to start with.

dwnusbaum committed:
[JENKINS-71715] Update help text and Javadoc based on review feedback, and filter out default provider in global config

Co-authored-by: Daniel Beck <daniel-beck@users.noreply.github.com>

dwnusbaum committed:
Merge branch 'master' into JENKINS-71715_first-approach

dwnusbaum commented on core/src/main/java/jenkins/console/ConsoleUrlProviderGlobalConfiguration.java:
```diff
@@ -0,0 +1,110 @@
+        return ExtensionList.lookupSingleton(ConsoleUrlProviderGlobalConfiguration.class);
+    }
+
+    public static List<? extends Descriptor<ConsoleUrlProvider>> getProvidersDescriptors() {
```
Let me check this, it probably doesn't need to be static.

dwnusbaum replied:
[88ea796](https://github.com/jenkinsci/jenkins/pull/8321/commits/88ea796c0c6d19cddbe758b28617c4b898db2b27)

dwnusbaum committed:
[JENKINS-71715] getProvidersDescriptors can be an instance method

MarkEWaite commented:
This PR is now ready for merge.  We will merge it after approximately 24 hours if there is no negative feedback.

/label ready-for-merge

dwnusbaum commented:
> > The current wording focuses on the URLs/links because that's what it does from a technical POV. From a user POV, it seems to me it's more about the chosen build log visualization

> Sure, we could change "Console URL Provider" to "Console Visualization Provider" at least in the user-facing text, and also the class names if desired.

@daniel-beck do you think I should make this change, or do you have any other suggestions?

daniel-beck commented:
> do you think I should make this change, or do you have any other suggestions?

Depends on whether you think this is easy followup, I guess (assuming you agree it would be a worthwhile change)? As I wrote, it's optional.

dwnusbaum commented:
Yes it is easy to change later (assuming we do not care about the class names). If you think "Console visualization provider" is better than "Console URL provider" though then I figure we might as well go ahead and change the text on the configuration pages here.

timja commented on core/src/main/java/jenkins/console/ConsoleUrlProviderGlobalConfiguration.java:
```diff
@@ -0,0 +1,110 @@
+        // For the global configuration, the default provider will always be consulted as a last resort, and since it
+        // handles all builds, there is no reason to ever select it explicitly.
+        return Jenkins.get().getDescriptorList(ConsoleUrlProvider.class).stream()
+                .filter(d -> !(d instanceof DefaultConsoleUrlProvider.DescriptorImpl))
```
it seems odd that this is filtered out here but not on the user property, adding it doesn't seem to make any difference on the user property, it will still always fall back to it

<img width="281" alt="image" src="https://github.com/jenkinsci/jenkins/assets/21194782/ce3b594b-5375-4047-9663-1b135a6e4e49">

daniel-beck replied:
In the global view, there's no difference between configuring none, and configuring Default. IMO it's confusing to have two different configurations that behave identically, so it makes sense to not be visible there.

In the user view, it's there to allow users to explicitly select Default no matter what the admin configures. It needs to exist because there's no other way for users to accomplish this.

timja replied:
Ah I see that makes sense.
